kernel/src/fs/overlayfs/fs.rs

```diff
@@ -0,0 +1,991 @@
+        upper.write_at(offset, reader)
+    }
+
+    pub fn write_direct_at(&self, offset: usize, reader: &mut VmReader) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        let upper = self.build_upper_recursively_if_needed()?;
+        upper.write_direct_at(offset, reader)
+    }
+
+    pub fn read_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        self.get_top_valid_inode().read_at(offset, writer)
+    }
+
+    pub fn read_direct_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        if self.type_ == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+        self.get_top_valid_inode().read_direct_at(offset, writer)
+    }
+
+    /// Returns the children objects in a unified view.
+    /// The object from the upper layer with the same name will mask the lower ones.
+    pub fn readdir_at(&self, offset: usize, visitor: &mut dyn DirentVisitor) -> Result<usize> {
+        if self.type_ != InodeType::Dir {
+            return_errno!(Errno::ENOTDIR);
+        }
+
+        let mut overlay_dir_visitor = OverlayDirMergedVisitor::new();
+        if let Some(upper) = self.upper() {
+            debug_assert!(upper.type_() == InodeType::Dir);
+            // The upper cannot be an opaque.
+            upper.readdir_at(offset, &mut overlay_dir_visitor)?;
+        }
+
+        for lower in self.lowers.iter() {
+            debug_assert!(lower.type_() == InodeType::Dir);
+            // The lower cannot be an opaque.
+            lower.readdir_at(offset, &mut overlay_dir_visitor)?;
+        }
+
+        let mut cnt = 0;
+        for (offset, (name, ino, type_)) in overlay_dir_visitor.as_merged_view() {
+            visitor.visit(name, *ino, *type_, *offset)?;
+            cnt += 1;
+        }
+        Ok(cnt)
+    }
+
+    /// Deletes the target file by creating a "whiteout" file from the upper layer.
+    /// The corresponding parent directories will be created also if they do not exist.
+    pub fn unlink(&self, name: &str) -> Result<()> {
+        let inode = self.lookup(name)?;
+        let target = inode.downcast_ref::<OverlayInode>().unwrap();
+        if target.type_() == InodeType::Dir {
+            return_errno!(Errno::EISDIR);
+        }
+
+        let mut upper_guard = self.upper.lock();
+        if upper_guard.is_none() {
+            drop(upper_guard);
+            self.build_upper_recursively_if_needed()?;
+            upper_guard = self.upper.lock();
+        }
+
+        let upper = upper_guard.as_ref().unwrap();
+        // The target file cannot be a whiteout.
+        if let Err(e) = upper.unlink(name) {
+            if e.error() != Errno::ENOENT {
+                return Err(e);
+            }
+        }
+
+        let whiteout = upper.create(name, InodeType::File, InodeMode::from_bits_truncate(0o644))?;
```

Comment:
I have two concerns.

First, my understanding is that whiteout files should be named specially, using a `.wh.` prefix in tradition. By checking the prefix, the `readdir` method can efficiently filter out deleted files. Otherwise, it has to access the extended attributes of the inode of each directory entry, which would trigger a lot of I/O.

Second, a whiteout file should be created only when we need to mark a file from the lower layers as deleted. But this unlink method creates a whiteout file even when the file only exists in the upper layer.

Reply:
https://github.com/torvalds/linux/blob/master/Documentation/filesystems/overlayfs.rst